File: python/leetcode/230.py
class Solution:
    def kthSmallest(self, root, k):
        self.k = k
        self.res = None
        self.help(root)
        return self.res

    def help(self, node):
        if not node: return
        self.helper(node.left)
        self.k -= 1
        if self.k == 0:
            self.res = node.val
            return
        self.helper(node.right)


def kthSmallest(root, k):
    stack = []
    while root or stack:
        while root:
            stack.append(root)
            root = root.left
        root = stack.pop()
        k -= 1
        if k == 0: return root.val
        root = root.right

File: cpp/leetcode/230.cpp
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        stack<TreeNode*> stk;
        while(!stk.empty() || root){
            if(root){
                stk.push(root);
                root = root->left;
            } else {
                TreeNode *node = stk.top(); stk.pop();
                k--;
                if(k == 0) return node->val;
                root = node->right;
            }
        }
        return -1;
    }
};


class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int res = 0;
        dfs(root, k, res);
        return res;
    }
    
    void dfs(TreeNode * root, int  &k, int &res){
        if(!root) return;
        dfs(root->left, k, res);
        k --;
        if(k == 0) res = root->val;
        dfs(root->right, k, res);
    }
};